### ISSUE TITLE
Collection-Register can be so slow that no result is returned?

### ISSUE THREAD

**Issue by GVogeler** [icaruseu/mom-ca#928]
When calling the index app in the private collection of IlluminierteUrkunden (https://www.monasterium.net/mom/IlluminierteUrkunden/collection-register?modus=mycollection) the "Register search" does not list any entries although a query modelled after https://github.com/icaruseu/mom-ca/blob/305738b87562c69f0760fb52b5e64382324168d7/my/XRX/src/mom/app/collection/collectionsearch.xqm#L67
```
xquery version "3.1";
declare namespace atom="http://www.w3.org/2005/Atom";
declare namespace cei="http://www.monasterium.net/NS/cei";
let $collection := collection('/db/mom-data/xrx.user/illuminierteurkunden%40gmail.com/metadata.charter/IlluminierteUrkunden')
let $node_name := 'cei:persName'
for $entries in distinct-values($collection//cei:text//*[name()=$node_name][starts-with(.,'A')])
return $entries
```
returns 223 entries in 2.6 seconds. Console does not return any information.

**Comment by GVogeler** [comment]
It seems that the `[name()=$node_name]` predicate is the expensive one. I would suggest to add a switch for the nodes in https://github.com/icaruseu/mom-ca/blob/305738b87562c69f0760fb52b5e64382324168d7/my/XRX/src/mom/app/collection/collectionsearch.xqm#L66:
```
let $nodes := switch ($node_name) 
   case "cei:persName" return $collection//cei:text//cei:persName
   case "cei:placeName" return $collection//cei:text//cei:placeName
   case "cei:index" return $collection//cei:text//cei:index
   default return $collection//cei:text//cei:*[name()=$node_name
for $entries in distinct-values($nodes) order by $entries return
```

**Comment by GVogeler** [comment]
... or er should consider to replace the `name()=$node_name` predicate by a `util:eval(concat('$collection//cei:text//',$node_name))` call?

**Comment by GVogeler** [comment]
additionally we should consider using `group-by` instead of `distinct-values()` as suggested by https://exist-db.org/exist/apps/doc/tuning#group-by

**Comment by GVogeler** [comment]
The options ranked by query time are:
slowest:
## 1
```
xquery version "3.1";
declare namespace atom="http://www.w3.org/2005/Atom";
declare namespace cei="http://www.monasterium.net/NS/cei";
declare namespace util="http://exist-db.org/xquery/util";
let $collection := collection('/db/mom-data/xrx.user/illuminierteurkunden%40gmail.com/metadata.charter/IlluminierteUrkunden')
let $node_name := 'cei:persName'
let $nodes := util:eval(concat('$collection//cei:text//',$node_name))
for $entries in $nodes 
    group by $entry := $entries/normalize-space()
    order by $entry
return $entry
```
much slower than:
## 2
```
xquery version "3.1";
declare namespace atom="http://www.w3.org/2005/Atom";
declare namespace cei="http://www.monasterium.net/NS/cei";
declare namespace util="http://exist-db.org/xquery/util";
let $collection := collection('/db/mom-data/xrx.user/illuminierteurkunden%40gmail.com/metadata.charter/IlluminierteUrkunden')
let $node_name := 'cei:persName'
for $entries in distinct-values($collection//cei:text//cei:*[name()=$node_name]/normalize-space())
    order by $entries
    return $entries
```
almost as slow as:
## 3
```
xquery version "3.1";
declare namespace atom="http://www.w3.org/2005/Atom";
declare namespace cei="http://www.monasterium.net/NS/cei";
declare namespace util="http://exist-db.org/xquery/util";
let $collection := collection('/db/mom-data/xrx.user/illuminierteurkunden%40gmail.com/metadata.charter/IlluminierteUrkunden')
let $node_name := 'cei:persName'
let $nodes := switch ($node_name) 
   case "cei:persName" return $collection//cei:text//cei:persName
   case "cei:placeName" return $collection//cei:text//cei:placeName
   case "cei:index" return $collection//cei:text//cei:index
   default return $collection//cei:text//cei:*[name()=$node_name]
for $entries in distinct-values($nodes/normalize-space())
    order by $entries
    return $entries
```
still slower - and much less flexible - than:
## 4
```
xquery version "3.1";
declare namespace atom="http://www.w3.org/2005/Atom";
declare namespace cei="http://www.monasterium.net/NS/cei";
declare namespace util="http://exist-db.org/xquery/util";
let $collection := collection('/db/mom-data/xrx.user/illuminierteurkunden%40gmail.com/metadata.charter/IlluminierteUrkunden')
let $node_name := 'cei:persName'
let $nodes := util:eval(concat('$collection//cei:text//',$node_name))
for $entries in distinct-values($nodes/normalize-space())
    order by $entries
    return $entries
```
So I would suggest solution 4.

**Comment by StephanMa** [comment]
deployed

**Comment by GVogeler** [comment]
closed with #931
